File: typescript/session2/td/heritageMultiple/SuccParInt.ts
import { AlgebreNatParInt, AlgebreNatParInt_defaut } from "./AlgebreNatParInt";
import { Nat, FabriqueNat } from "../../../session1/td/naturels";
import { ZeroParInt } from "./ZeroParInt";
import { EtatSucc } from "./EtatSucc";
import { heriterInterfacesConcretes } from "../../../heritageMultiple/heritageInterfacesconcretes";

export class SuccParInt extends EtatSucc implements AlgebreNatParInt {
    public static readonly FAB: FabriqueNat<Nat> = new SuccParInt(new ZeroParInt());

    constructor(_predecesseur : Nat) {
        super(_predecesseur)
    }

    creerZero() : Nat {
        return new ZeroParInt;
    }

    creerSuccesseur(predecesseur : Nat) : Nat {
        return new SuccParInt(predecesseur);
    }

    estEgal(x : Nat) {
        return x.representation() === this.representation()
    }

    representation() : string {
        return String(this.val())
    }

    somme: (x: Nat) => Nat;
    produit: (x: Nat) => Nat;
	div: (x: Nat) => Nat;
	modulo: (x: Nat) => Nat;
	zero: () => Nat;
    un: () => Nat;
    val: () => number;
    estNul: () => boolean;
    predecesseur: () => Nat;
    taille: () => number;
    chiffre: (i: number) => number;
    creerNatAvecValeur: (x: number) => Nat;
    creerNatAvecRepresentation: (repDecimal: string) => Nat;
}

heriterInterfacesConcretes(SuccParInt, [EtatSucc, AlgebreNatParInt_defaut])

File: java/cours/src/session2/td/heritageMultiple/SuccParInt.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public class SuccParInt extends EtatSucc implements AlgebreNatParInt {

	public SuccParInt(Nat val) {
		super(val);
	}
	
	public Nat creerZero() {
		return new ZeroParInt();
	}
	
	public Nat creerSuccesseur(Nat predecesseur) {
		return new SuccParInt(predecesseur);
	}
	
	public String toString() {
		return String.valueOf(super.val());
	}
	
	public boolean equals(Object obj) {
		if (!(obj instanceof Nat))
			return false;
		Nat x = (Nat) obj;
		return this.val() == x.val();
	}
}
